// @smc/tunnel-service/test/log.js
if (typeof process.env.LOG_LEVEL == 'undefined') process.env.LOG_LEVEL = 'info';

var util = require('util');

var log = function(level, argz) {
  var args = Array.prototype.slice.call(argz);
  var message = util.format.apply(null, args);
  console.log(util.format('[%s]', level), message);
}

module.exports = {
  fatal: function() {
    if (process.env.LOG_LEVEL != 'off' && ['all', 'trace', 'debug', 'info', 'warn', 'error', 'fatal'].indexOf(process.env.LOG_LEVEL) >= 0)
      log('fatal', arguments);
  },
  error: function() {
    if (process.env.LOG_LEVEL != 'off' && ['all', 'trace', 'debug', 'info', 'warn', 'error'].indexOf(process.env.LOG_LEVEL) >= 0)
      log('error', arguments);
  },
  warn: function() {
    if (process.env.LOG_LEVEL != 'off' && ['all', 'trace', 'debug', 'info', 'warn'].indexOf(process.env.LOG_LEVEL) >= 0)
      log('warn', arguments);
  },
  info: function() {
    if (process.env.LOG_LEVEL != 'off' && ['all', 'trace', 'debug', 'info'].indexOf(process.env.LOG_LEVEL) >= 0)
      log('info', arguments);
  },
  $$DEBUG: function() {
    if (process.env.LOG_LEVEL != 'off' && ['all', 'trace', 'debug'].indexOf(process.env.LOG_LEVEL) >= 0)
      log('debug', arguments);
  },
  $$TRACE: function() {
    if (process.env.LOG_LEVEL != 'off' && ['all', 'trace'].indexOf(process.env.LOG_LEVEL) >= 0)
      log('trace', arguments);
  },
}
